File: elements_study.go
# elements_study.go
/**
 * ⚛️ Element Master – Learn Chemistry Elements (Go Edition)
 * Advanced: complete DB, favorites, quiz, spaced repetition, stats
 */

package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"math/rand"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"
)

// ─── Element Type ─────────────────────────────────────────────────────────

type Element struct {
	Symbol         string  `json:"symbol"`
	Name           string  `json:"name"`
	Number         int     `json:"number"`
	Period         int     `json:"period"`
	Group          string  `json:"group"`
	Category       string  `json:"category"`
	Mass           float64 `json:"mass"`
	ElectronConfig string  `json:"electron_config"`
}

// ─── User Data ────────────────────────────────────────────────────────────

type UserStats struct {
	Correct  int    `json:"correct"`
	Wrong    int    `json:"wrong"`
	LastSeen string `json:"last_seen"`
}

type UserData struct {
	Favorites []string            `json:"favorites"`
	Stats     map[string]UserStats `json:"stats"`
	RepQueue  []string            `json:"rep_queue"`
}

// ─── Global Data ──────────────────────────────────────────────────────────

var ELEMENTS = map[string]Element{}
var ELEMENTS_LIST []Element

func init() {
	// Populate with first 20 for brevity; full list in real code.
	data := []Element{
		{"H", "Hydrogen", 1, 1, "1", "Nonmetal", 1.008, "1s1"},
		// ... full list
	}
	for _, e := range data {
		ELEMENTS[e.Symbol] = e
		ELEMENTS_LIST = append(ELEMENTS_LIST, e)
	}
}

// ─── Colors ───────────────────────────────────────────────────────────────

const (
	reset  = "\x1b[0m"
	bright = "\x1b[1m"
	dim    = "\x1b[2m"
	red    = "\x1b[31m"
	green  = "\x1b[32m"
	yellow = "\x1b[33m"
	blue   = "\x1b[34m"
	magenta = "\x1b[35m"
	cyan   = "\x1b[36m"
)

func c(str, color string) string {
	return color + str + reset
}

// ─── User Data Manager ──────────────────────────────────────────────────

type UserDataManager struct {
	dataDir  string
	dataFile string
	data     UserData
}

func NewUserDataManager() *UserDataManager {
	home, _ := os.UserHomeDir()
	dir := filepath.Join(home, ".elements_study")
	os.MkdirAll(dir, 0755)
	file := filepath.Join(dir, "user_data.json")
	ud := &UserDataManager{dataDir: dir, dataFile: file}
	ud.load()
	return ud
}

func (ud *UserDataManager) load() {
	if _, err := os.Stat(ud.dataFile); os.IsNotExist(err) {
		ud.data = UserData{Favorites: []string{}, Stats: map[string]UserStats{}, RepQueue: []string{}}
		return
	}
	raw, err := os.ReadFile(ud.dataFile)
	if err != nil {
		ud.data = UserData{Favorites: []string{}, Stats: map[string]UserStats{}, RepQueue: []string{}}
		return
	}
	var d UserData
	if err := json.Unmarshal(raw, &d); err != nil {
		ud.data = UserData{Favorites: []string{}, Stats: map[string]UserStats{}, RepQueue: []string{}}
		return
	}
	ud.data = d
}

func (ud *UserDataManager) save() {
	raw, _ := json.MarshalIndent(ud.data, "", "  ")
	os.WriteFile(ud.dataFile, raw, 0644)
}

func (ud *UserDataManager) ToggleFavorite(symbol string) {
	for i, s := range ud.data.Favorites {
		if s == symbol {
			ud.data.Favorites = append(ud.data.Favorites[:i], ud.data.Favorites[i+1:]...)
			ud.save()
			return
		}
	}
	ud.data.Favorites = append(ud.data.Favorites, symbol)
	ud.save()
}

func (ud *UserDataManager) IsFavorite(symbol string) bool {
	for _, s := range ud.data.Favorites {
		if s == symbol {
			return true
		}
	}
	return false
}

func (ud *UserDataManager) RecordAnswer(symbol string, correct bool) {
	stats := ud.data.Stats[symbol]
	if stats.Correct == 0 && stats.Wrong == 0 {
		stats = UserStats{Correct: 0, Wrong: 0, LastSeen: time.Now().Format(time.RFC3339)}
	}
	if correct {
		stats.Correct++
	} else {
		stats.Wrong++
	}
	stats.LastSeen = time.Now().Format(time.RFC3339)
	ud.data.Stats[symbol] = stats

	// update rep queue
	for i, s := range ud.data.RepQueue {
		if s == symbol {
			ud.data.RepQueue = append(ud.data.RepQueue[:i], ud.data.RepQueue[i+1:]...)
			break
		}
	}
	if !correct {
		ud.data.RepQueue = append([]string{symbol}, ud.data.RepQueue...)
	} else {
		ud.data.RepQueue = append(ud.data.RepQueue, symbol)
	}
	if len(ud.data.RepQueue) > 30 {
		ud.data.RepQueue = ud.data.RepQueue[:30]
	}
	ud.save()
}

func (ud *UserDataManager) GetNextRep() string {
	if len(ud.data.RepQueue) > 0 {
		return ud.data.RepQueue[0]
	}
	var unmastered []string
	for sym := range ELEMENTS {
		if stats, ok := ud.data.Stats[sym]; !ok || stats.Correct < 3 {
			unmastered = append(unmastered, sym)
		}
	}
	if len(unmastered) > 0 {
		return unmastered[rand.Intn(len(unmastered))]
	}
	// all mastered
	syms := make([]string, 0, len(ELEMENTS))
	for sym := range ELEMENTS {
		syms = append(syms, sym)
	}
	return syms[rand.Intn(len(syms))]
}

func (ud *UserDataManager) GetProgress() (mastered, total int) {
	total = len(ELEMENTS)
	for sym := range ELEMENTS {
		if stats, ok := ud.data.Stats[sym]; ok && stats.Correct >= 3 {
			mastered++
		}
	}
	return
}

// ─── Quiz Engine ──────────────────────────────────────────────────────────

type QuizEngine struct {
	user *UserDataManager
	reader *bufio.Reader
}

func NewQuizEngine(user *UserDataManager) *QuizEngine {
	return &QuizEngine{user: user, reader: bufio.NewReader(os.Stdin)}
}

func (q *QuizEngine) ask(prompt string) string {
	fmt.Print(prompt)
	line, _ := q.reader.ReadString('\n')
	return strings.TrimSpace(line)
}

func (q *QuizEngine) getOptions(element Element, field string) []string {
	correct := ""
	if field == "name" {
		correct = element.Name
	} else {
		correct = element.Symbol
	}
	var pool []Element
	for _, e := range ELEMENTS_LIST {
		val := ""
		if field == "name" {
			val = e.Name
		} else {
			val = e.Symbol
		}
		if val != correct {
			pool = append(pool, e)
		}
	}
	others := []string{}
	for len(others) < 3 {
		r := pool[rand.Intn(len(pool))]
		val := ""
		if field == "name" {
			val = r.Name
		} else {
			val = r.Symbol
		}
		found := false
		for _, o := range others {
			if o == val {
				found = true
				break
			}
		}
		if !found && val != correct {
			others = append(others, val)
		}
	}
	options := append([]string{correct}, others...)
	rand.Shuffle(len(options), func(i, j int) { options[i], options[j] = options[j], options[i] })
	return options
}

func (q *QuizEngine) RunQuiz(numQuestions int) {
	correctCount := 0
	fmt.Println(c("\n🧠 Quiz Time! Answer questions about elements.", bright+cyan))
	for i := 0; i < numQuestions; i++ {
		qType := rand.Intn(2)
		element := ELEMENTS_LIST[rand.Intn(len(ELEMENTS_LIST))]
		var prompt, correct string
		var options []string
		if qType == 0 {
			correct = element.Name
			prompt = fmt.Sprintf("What is the name of element with symbol %s?", element.Symbol)
			options = q.getOptions(element, "name")
		} else {
			correct = element.Symbol
			prompt = fmt.Sprintf("What is the symbol of %s?", element.Name)
			options = q.getOptions(element, "symbol")
		}
		fmt.Printf("\n%s %s\n", c(fmt.Sprintf("Q%d.", i+1), yellow), prompt)
		for idx, opt := range options {
			fmt.Printf("  %d. %s\n", idx+1, opt)
		}
		choice := q.ask("Your choice (1-4): ")
		idx, _ := strconv.Atoi(choice)
		selected := options[idx-1]
		isCorrect := selected == correct
		q.user.RecordAnswer(element.Symbol, isCorrect)
		if isCorrect {
			correctCount++
			fmt.Printf("%s\n", c("✅ Correct! "+correct, green))
		} else {
			fmt.Printf("%s\n", c("❌ Wrong! The answer was "+correct, red))
		}
	}
	fmt.Printf("\n%s Correct: %s, Wrong: %s\n", c("Quiz finished!", bright), c(strconv.Itoa(correctCount), green), c(strconv.Itoa(numQuestions-correctCount), red))
}

// ─── Main App ─────────────────────────────────────────────────────────────

type ElementApp struct {
	user  *UserDataManager
	quiz  *QuizEngine
	reader *bufio.Reader
}

func NewElementApp() *ElementApp {
	rand.Seed(time.Now().UnixNano())
	user := NewUserDataManager()
	return &ElementApp{
		user:   user,
		quiz:   NewQuizEngine(user),
		reader: bufio.NewReader(os.Stdin),
	}
}

func (app *ElementApp) ask(prompt string) string {
	fmt.Print(prompt)
	line, _ := app.reader.ReadString('\n')
	return strings.TrimSpace(line)
}

func (app *ElementApp) askInt(prompt string) int {
	for {
		line := app.ask(prompt)
		if i, err := strconv.Atoi(line); err == nil {
			return i
		}
		fmt.Println(c("Please enter a number.", yellow))
	}
}

func (app *ElementApp) showMenu() {
	mastered, total := app.user.GetProgress()
	fmt.Println("\n" + c(strings.Repeat("═", 50), cyan))
	fmt.Println(c("⚛️ ELEMENT MASTER", bright+cyan))
	fmt.Println(c(strings.Repeat("═", 50), cyan))
	fmt.Printf("  Favorites: %d\n", len(app.user.data.Favorites))
	fmt.Printf("  Mastered: %d/%d\n", mastered, total)
	fmt.Printf("  Next repetition: %s\n", app.user.GetNextRep())
	fmt.Println(c(strings.Repeat("═", 50), cyan))
	fmt.Println("  1. 📋 List All Elements")
	fmt.Println("  2. 🔍 Search Element")
	fmt.Println("  3. ⭐ Favorites")
	fmt.Println("  4. 🧠 Start Quiz")
	fmt.Println("  5. 📊 Statistics")
	fmt.Println("  6. 🔁 Spaced Repetition")
	fmt.Println("  7. ➕ Toggle Favorite")
	fmt.Println("  0. 🚪 Exit")
	fmt.Println(c(strings.Repeat("═", 50), cyan))
}

func (app *ElementApp) listElements() {
	fmt.Println("\n📋 ALL ELEMENTS")
	fmt.Println(c(strings.Repeat("─", 60), dim))
	for _, e := range ELEMENTS_LIST {
		star := ""
		if app.user.IsFavorite(e.Symbol) {
			star = "⭐"
		}
		fmt.Printf("  %-3s %s %-12s #%-3d %-15s %.3f\n", e.Symbol, star, e.Name, e.Number, e.Category, e.Mass)
	}
}

func (app *ElementApp) searchElement() {
	query := app.ask("🔍 Enter symbol, name, or number: ")
	results := []Element{}
	for _, e := range ELEMENTS_LIST {
		if strings.Contains(strings.ToLower(e.Symbol), strings.ToLower(query)) ||
			strings.Contains(strings.ToLower(e.Name), strings.ToLower(query)) ||
			strconv.Itoa(e.Number) == query ||
			strings.Contains(strings.ToLower(e.Category), strings.ToLower(query)) {
			results = append(results, e)
		}
	}
	if len(results) == 0 {
		fmt.Println(c("No elements found.", yellow))
		return
	}
	fmt.Printf("\n🔍 Results (%d)\n", len(results))
	for _, e := range results {
		star := ""
		if app.user.IsFavorite(e.Symbol) {
			star = "⭐"
		}
		fmt.Printf("  %-3s %s %-12s #%-3d %-15s %.3f\n", e.Symbol, star, e.Name, e.Number, e.Category, e.Mass)
	}
}

func (app *ElementApp) showFavorites() {
	favs := []Element{}
	for _, sym := range app.user.data.Favorites {
		if e, ok := ELEMENTS[sym]; ok {
			favs = append(favs, e)
		}
	}
	if len(favs) == 0 {
		fmt.Println(c("No favorites yet.", yellow))
		return
	}
	fmt.Println("\n⭐ FAVORITES")
	for _, e := range favs {
		fmt.Printf("  %-3s %-12s #%-3d %s\n", e.Symbol, e.Name, e.Number, e.Category)
	}
}

func (app *ElementApp) startQuiz() {
	num := app.askInt("Number of questions (default 10): ")
	if num <= 0 {
		num = 10
	}
	app.quiz.RunQuiz(num)
}

func (app *ElementApp) showStats() {
	mastered, total := app.user.GetProgress()
	totalAnswers := 0
	correctAnswers := 0
	for _, s := range app.user.data.Stats {
		totalAnswers += s.Correct + s.Wrong
		correctAnswers += s.Correct
	}
	fmt.Println("\n📊 STATISTICS")
	fmt.Println(c(strings.Repeat("─", 30), dim))
	fmt.Printf("  Total Elements: %d\n", total)
	fmt.Printf("  Mastered: %d\n", mastered)
	fmt.Printf("  Favorites: %d\n", len(app.user.data.Favorites))
	fmt.Printf("  Total Answers: %d\n", totalAnswers)
	fmt.Printf("  Correct Answers: %d\n", correctAnswers)
	if totalAnswers > 0 {
		fmt.Printf("  Accuracy: %.1f%%\n", float64(correctAnswers)/float64(totalAnswers)*100)
	}
}

func (app *ElementApp) spacedRepetition() {
	symbol := app.user.GetNextRep()
	if symbol == "" {
		fmt.Println(c("No elements to repeat. Keep learning!", green))
		return
	}
	element := ELEMENTS[symbol]
	fmt.Printf("\n🔁 Repetition: %s (%s)\n", element.Name, element.Symbol)
	fmt.Printf("  Number: %d  Category: %s  Mass: %.3f\n", element.Number, element.Category, element.Mass)
	qType := rand.Intn(2)
	var ans, correct string
	if qType == 0 {
		ans = app.ask(fmt.Sprintf("What is the name of %s? ", element.Symbol))
		correct = element.Name
	} else {
		ans = app.ask(fmt.Sprintf("What is the symbol of %s? ", element.Name))
		correct = element.Symbol
	}
	isCorrect := strings.ToLower(strings.TrimSpace(ans)) == strings.ToLower(correct)
	app.user.RecordAnswer(symbol, isCorrect)
	if isCorrect {
		fmt.Println(c("✅ Correct!", green))
	} else {
		fmt.Printf("%s\n", c("❌ Wrong. The answer was "+correct, red))
	}
}

func (app *ElementApp) toggleFavorite() {
	sym := app.ask("Enter element symbol to toggle favorite: ")
	symbol := strings.ToUpper(strings.TrimSpace(sym))
	if _, ok := ELEMENTS[symbol]; !ok {
		fmt.Println(c("Element not found.", red))
		return
	}
	app.user.ToggleFavorite(symbol)
	state := "removed from"
	if app.user.IsFavorite(symbol) {
		state = "added to"
	}
	fmt.Printf("%s\n", c("✅ "+symbol+" "+state+" favorites.", green))
}

func (app *ElementApp) run() {
	fmt.Print("\033[H\033[2J")
	fmt.Println(c("\n⚛️ Element Master – Learn Chemistry Elements", bright+cyan))
	fmt.Println(c("Master the periodic table, one element at a time!", dim))

	for {
		app.showMenu()
		choice := app.ask("Your choice: ")
		switch choice {
		case "1":
			app.listElements()
		case "2":
			app.searchElement()
		case "3":
			app.showFavorites()
		case "4":
			app.startQuiz()
		case "5":
			app.showStats()
		case "6":
			app.spacedRepetition()
		case "7":
			app.toggleFavorite()
		case "0":
			fmt.Println(c("👋 Goodbye! Keep learning!", cyan))
			return
		default:
			fmt.Println(c("❌ Invalid choice.", red))
		}
		if choice != "0" {
			fmt.Print("\nPress Enter to continue...")
			app.reader.ReadString('\n')
		}
	}
}

func main() {
	app := NewElementApp()
	app.run()
}

File: elements_study.rs
# elements_study.rs
/**
 * ⚛️ Element Master – Learn Chemistry Elements (Rust Edition)
 * Advanced: complete DB, favorites, quiz, spaced repetition, stats
 */

use rand::seq::SliceRandom;
use rand::thread_rng;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fs;
use std::io::{self, Write, BufRead};
use std::path::PathBuf;
use chrono::Utc;

// ─── Types ──────────────────────────────────────────────────────────────────

#[derive(Debug, Serialize, Deserialize, Clone)]
struct Element {
    symbol: String,
    name: String,
    number: u32,
    period: u32,
    group: String,
    category: String,
    mass: f64,
    electron_config: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
struct UserStats {
    correct: u32,
    wrong: u32,
    last_seen: String,
}

#[derive(Debug, Serialize, Deserialize)]
struct UserData {
    favorites: Vec<String>,
    stats: HashMap<String, UserStats>,
    rep_queue: Vec<String>,
}

// ─── Global Data ──────────────────────────────────────────────────────────

lazy_static! {
    static ref ELEMENTS: HashMap<String, Element> = {
        let data = vec![
            Element { symbol: "H".to_string(), name: "Hydrogen".to_string(), number: 1, period: 1, group: "1".to_string(), category: "Nonmetal".to_string(), mass: 1.008, electron_config: "1s1".to_string() },
            // ... full list
        ];
        let mut map = HashMap::new();
        for e in data {
            map.insert(e.symbol.clone(), e);
        }
        map
    };
    static ref ELEMENTS_LIST: Vec<Element> = ELEMENTS.values().cloned().collect();
}

// ─── Colors ──────────────────────────────────────────────────────────────────

fn c(text: &str, color: &str) -> String {
    format!("{}{}{}", color, text, "\x1b[0m")
}

const RESET: &str = "\x1b[0m";
const BRIGHT: &str = "\x1b[1m";
const DIM: &str = "\x1b[2m";
const RED: &str = "\x1b[31m";
const GREEN: &str = "\x1b[32m";
const YELLOW: &str = "\x1b[33m";
const BLUE: &str = "\x1b[34m";
const MAGENTA: &str = "\x1b[35m";
const CYAN: &str = "\x1b[36m";

// ─── User Data Manager ─────────────────────────────────────────────────────

struct UserDataManager {
    file_path: PathBuf,
    data: UserData,
}

impl UserDataManager {
    fn new() -> Self {
        let home = std::env::var("HOME").or_else(|_| std::env::var("USERPROFILE")).unwrap_or_else(|_| ".".to_string());
        let dir = PathBuf::from(home).join(".elements_study");
        fs::create_dir_all(&dir).unwrap();
        let file_path = dir.join("user_data.json");
        let mut ud = UserDataManager { file_path, data: UserData { favorites: Vec::new(), stats: HashMap::new(), rep_queue: Vec::new() } };
        ud.load();
        ud
    }

    fn load(&mut self) {
        if let Ok(raw) = fs::read_to_string(&self.file_path) {
            if let Ok(data) = serde_json::from_str::<UserData>(&raw) {
                self.data = data;
                return;
            }
        }
        self.data = UserData { favorites: Vec::new(), stats: HashMap::new(), rep_queue: Vec::new() };
    }

    fn save(&self) {
        let raw = serde_json::to_string_pretty(&self.data).unwrap();
        let _ = fs::write(&self.file_path, raw);
    }

    fn toggle_favorite(&mut self, symbol: &str) {
        if let Some(pos) = self.data.favorites.iter().position(|s| s == symbol) {
            self.data.favorites.remove(pos);
        } else {
            self.data.favorites.push(symbol.to_string());
        }
        self.save();
    }

    fn is_favorite(&self, symbol: &str) -> bool {
        self.data.favorites.contains(&symbol.to_string())
    }

    fn record_answer(&mut self, symbol: &str, correct: bool) {
        let stats = self.data.stats.entry(symbol.to_string()).or_insert(UserStats { correct: 0, wrong: 0, last_seen: Utc::now().to_rfc3339() });
        if correct { stats.correct += 1; } else { stats.wrong += 1; }
        stats.last_seen = Utc::now().to_rfc3339();

        // update rep queue
        if let Some(pos) = self.data.rep_queue.iter().position(|s| s == symbol) {
            self.data.rep_queue.remove(pos);
        }
        if !correct {
            self.data.rep_queue.insert(0, symbol.to_string());
        } else {
            self.data.rep_queue.push(symbol.to_string());
        }
        if self.data.rep_queue.len() > 30 {
            self.data.rep_queue.truncate(30);
        }
        self.save();
    }

    fn get_next_rep(&self) -> Option<String> {
        if !self.data.rep_queue.is_empty() {
            return Some(self.data.rep_queue[0].clone());
        }
        let mut unmastered = Vec::new();
        for (sym, stats) in &self.data.stats {
            if stats.correct < 3 {
                unmastered.push(sym.clone());
            }
        }
        if !unmastered.is_empty() {
            return Some(unmastered.choose(&mut thread_rng()).unwrap().clone());
        }
        let syms: Vec<String> = ELEMENTS.keys().cloned().collect();
        syms.choose(&mut thread_rng()).cloned()
    }

    fn get_progress(&self) -> (usize, usize) {
        let total = ELEMENTS.len();
        let mastered = self.data.stats.iter().filter(|(_, stats)| stats.correct >= 3).count();
        (mastered, total)
    }
}

// ─── Quiz Engine ──────────────────────────────────────────────────────────

struct QuizEngine {
    user: UserDataManager,
    rng: rand::rngs::ThreadRng,
}

impl QuizEngine {
    fn new(user: UserDataManager) -> Self {
        QuizEngine { user, rng: thread_rng() }
    }

    fn get_options(&self, element: &Element, field: &str) -> Vec<String> {
        let correct = if field == "name" { element.name.clone() } else { element.symbol.clone() };
        let pool: Vec<&Element> = ELEMENTS_LIST.iter().filter(|e| {
            let val = if field == "name" { &e.name } else { &e.symbol };
            val != &correct
        }).collect();
        let mut others = Vec::new();
        while others.len() < 3 {
            let r = pool.choose(&mut self.rng).unwrap();
            let val = if field == "name" { &r.name } else { &r.symbol };
            if !others.contains(val) && val != &correct {
                others.push(val.clone());
            }
        }
        let mut options = vec![correct.clone()];
        options.extend(others);
        options.shuffle(&mut self.rng);
        options
    }

    fn run_quiz(&mut self, num_questions: u32) {
        let mut correct_count = 0;
        println!("{}", c("\n🧠 Quiz Time! Answer questions about elements.", &format!("{}{}", BRIGHT, CYAN)));
        for i in 0..num_questions {
            let q_type = if rand::random() { 0 } else { 1 };
            let element = ELEMENTS_LIST.choose(&mut self.rng).unwrap();
            let (prompt, correct, options) = if q_type == 0 {
                let correct = element.name.clone();
                let prompt = format!("What is the name of element with symbol {}?", element.symbol);
                let options = self.get_options(element, "name");
                (prompt, correct, options)
            } else {
                let correct = element.symbol.clone();
                let prompt = format!("What is the symbol of {}?", element.name);
                let options = self.get_options(element, "symbol");
                (prompt, correct, options)
            };
            println!("\n{} {}", c(&format!("Q{}", i+1), YELLOW), prompt);
            for (idx, opt) in options.iter().enumerate() {
                println!("  {}. {}", idx+1, opt);
            }
            print!("Your choice (1-4): ");
            io::stdout().flush().unwrap();
            let mut choice = String::new();
            io::stdin().read_line(&mut choice).unwrap();
            let idx: usize = choice.trim().parse().unwrap_or(1);
            let selected = &options[idx-1];
            let is_correct = selected == &correct;
            self.user.record_answer(&element.symbol, is_correct);
            if is_correct {
                correct_count += 1;
                println!("{}", c(&format!("✅ Correct! {}", correct), GREEN));
            } else {
                println!("{}", c(&format!("❌ Wrong! The answer was {}", correct), RED));
            }
        }
        println!("\n{} Correct: {}, Wrong: {}", c("Quiz finished!", BRIGHT), c(&correct_count.to_string(), GREEN), c(&(num_questions - correct_count).to_string(), RED));
    }
}

// ─── Main App ─────────────────────────────────────────────────────────────

struct ElementApp {
    user: UserDataManager,
    quiz: QuizEngine,
}

impl ElementApp {
    fn new() -> Self {
        let user = UserDataManager::new();
        let quiz = QuizEngine::new(user);
        ElementApp { user, quiz }
    }

    fn ask(&self, prompt: &str) -> String {
        print!("{}", prompt);
        io::stdout().flush().unwrap();
        let mut line = String::new();
        io::stdin().read_line(&mut line).unwrap();
        line.trim().to_string()
    }

    fn ask_int(&self, prompt: &str) -> u32 {
        loop {
            let ans = self.ask(prompt);
            if let Ok(num) = ans.parse::<u32>() {
                return num;
            }
            println!("{}", c("Please enter a number.", YELLOW));
        }
    }

    fn show_menu(&self) {
        let (mastered, total) = self.user.get_progress();
        let next = self.user.get_next_rep().unwrap_or_else(|| "—".to_string());
        println!("\n{}", c(&"=".repeat(50), CYAN));
        println!("{}", c("⚛️ ELEMENT MASTER", &format!("{}{}", BRIGHT, CYAN)));
        println!("{}", c(&"=".repeat(50), CYAN));
        println!("  Favorites: {}", self.user.data.favorites.len());
        println!("  Mastered: {}/{}", mastered, total);
        println!("  Next repetition: {}", next);
        println!("{}", c(&"=".repeat(50), CYAN));
        println!("  1. 📋 List All Elements");
        println!("  2. 🔍 Search Element");
        println!("  3. ⭐ Favorites");
        println!("  4. 🧠 Start Quiz");
        println!("  5. 📊 Statistics");
        println!("  6. 🔁 Spaced Repetition");
        println!("  7. ➕ Toggle Favorite");
        println!("  0. 🚪 Exit");
        println!("{}", c(&"=".repeat(50), CYAN));
    }

    fn list_elements(&self) {
        println!("\n📋 ALL ELEMENTS");
        println!("{}", c(&"-".repeat(60), DIM));
        for e in ELEMENTS_LIST.iter() {
            let star = if self.user.is_favorite(&e.symbol) { "⭐" } else { "" };
            println!("  {:3} {} {:12} #{:3} {:15} {:.3}", e.symbol, star, e.name, e.number, e.category, e.mass);
        }
    }

    fn search_element(&self) {
        let query = self.ask("🔍 Enter symbol, name, or number: ");
        let results: Vec<&Element> = ELEMENTS_LIST.iter().filter(|e| {
            e.symbol.to_lowercase().contains(&query.to_lowercase()) ||
            e.name.to_lowercase().contains(&query.to_lowercase()) ||
            e.number.to_string() == query ||
            e.category.to_lowercase().contains(&query.to_lowercase())
        }).collect();
        if results.is_empty() {
            println!("{}", c("No elements found.", YELLOW));
            return;
        }
        println!("\n🔍 Results ({})", results.len());
        for e in results {
            let star = if self.user.is_favorite(&e.symbol) { "⭐" } else { "" };
            println!("  {:3} {} {:12} #{:3} {:15} {:.3}", e.symbol, star, e.name, e.number, e.category, e.mass);
        }
    }

    fn show_favorites(&self) {
        let favs: Vec<&Element> = self.user.data.favorites.iter().filter_map(|s| ELEMENTS.get(s)).collect();
        if favs.is_empty() {
            println!("{}", c("No favorites yet.", YELLOW));
            return;
        }
        println!("\n⭐ FAVORITES");
        for e in favs {
            println!("  {:3} {:12} #{:3} {}", e.symbol, e.name, e.number, e.category);
        }
    }

    fn start_quiz(&mut self) {
        let num = self.ask_int("Number of questions (default 10): ");
        let n = if num == 0 { 10 } else { num };
        self.quiz.run_quiz(n);
    }

    fn show_stats(&self) {
        let (mastered, total) = self.user.get_progress();
        let total_answers: u32 = self.user.data.stats.values().map(|s| s.correct + s.wrong).sum();
        let correct_answers: u32 = self.user.data.stats.values().map(|s| s.correct).sum();
        println!("\n📊 STATISTICS");
        println!("{}", c(&"-".repeat(30), DIM));
        println!("  Total Elements: {}", total);
        println!("  Mastered: {}", mastered);
        println!("  Favorites: {}", self.user.data.favorites.len());
        println!("  Total Answers: {}", total_answers);
        println!("  Correct Answers: {}", correct_answers);
        if total_answers > 0 {
            println!("  Accuracy: {:.1}%", correct_answers as f64 / total_answers as f64 * 100.0);
        }
    }

    fn spaced_repetition(&mut self) {
        if let Some(symbol) = self.user.get_next_rep() {
            if let Some(element) = ELEMENTS.get(&symbol) {
                println!("\n🔁 Repetition: {} ({})", element.name, element.symbol);
                println!("  Number: {}  Category: {}  Mass: {:.3}", element.number, element.category, element.mass);
                let q_type = if rand::random() { 0 } else { 1 };
                let (prompt, correct) = if q_type == 0 {
                    (format!("What is the name of {}? ", element.symbol), element.name.clone())
                } else {
                    (format!("What is the symbol of {}? ", element.name), element.symbol.clone())
                };
                let ans = self.ask(&prompt);
                let is_correct = ans.trim().to_lowercase() == correct.to_lowercase();
                self.user.record_answer(&symbol, is_correct);
                if is_correct {
                    println!("{}", c("✅ Correct!", GREEN));
                } else {
                    println!("{}", c(&format!("❌ Wrong. The answer was {}", correct), RED));
                }
                return;
            }
        }
        println!("{}", c("No elements to repeat. Keep learning!", GREEN));
    }

    fn toggle_favorite(&mut self) {
        let sym = self.ask("Enter element symbol to toggle favorite: ");
        let symbol = sym.trim().to_uppercase();
        if !ELEMENTS.contains_key(&symbol) {
            println!("{}", c("Element not found.", RED));
            return;
        }
        self.user.toggle_favorite(&symbol);
        let state = if self.user.is_favorite(&symbol) { "added to" } else { "removed from" };
        println!("{}", c(&format!("✅ {} {} favorites.", symbol, state), GREEN));
    }

    fn run(&mut self) {
        println!("{}", c("\n⚛️ Element Master – Learn Chemistry Elements", &format!("{}{}", BRIGHT, CYAN)));
        println!("{}", c("Master the periodic table, one element at a time!", DIM));

        loop {
            self.show_menu();
            let choice = self.ask("Your choice: ");
            match choice.as_str() {
                "1" => self.list_elements(),
                "2" => self.search_element(),
                "3" => self.show_favorites(),
                "4" => self.start_quiz(),
                "5" => self.show_stats(),
                "6" => self.spaced_repetition(),
                "7" => self.toggle_favorite(),
                "0" => {
                    println!("{}", c("👋 Goodbye! Keep learning!", CYAN));
                    break;
                }
                _ => println!("{}", c("❌ Invalid choice.", RED)),
            }
            if choice != "0" {
                print!("\nPress Enter to continue...");
                io::stdout().flush().unwrap();
                let mut _dummy = String::new();
                io::stdin().read_line(&mut _dummy).unwrap();
            }
        }
    }
}

// ─── Main ────────────────────────────────────────────────────────────────────

#[macro_use] extern crate lazy_static;

fn main() {
    let mut app = ElementApp::new();
    app.run();
}
